doc/design/effects_mutability.md

```diff
@@ -0,0 +1,266 @@
+# Effects and mutability
+
+Hemlock statically determines all effects and mutability transitively associated
+with a type, whether direct or indirect.  Internal effects and mutability are
+prohibited in global and message values, though external effects (i.e. outside
+the Hemlock execution environment) are allowed.  Internal effects are lexically
+bound to a particular actor's local heap, and therefore cannot be supported in
+the global heap.  These restrictions are evaluated at compile time, which means
+that the programming model is unaffected by type erasure.  The absence of
+dynamic validation overhead enables fast message passing, as well as fully
+automated migration of compatible long-lived values to the global heap.
+
+Non-parametric types have fixed effects and mutability, whereas parametric types
+also have transitive effects and/or mutability that allow differing
+parametrizations to be any combination of effectless/effectful and/or
+immutable/mutable.  For example, `'a array` is directly immutable, but it may be
+transitively mutable depending on the particulars of the type supplied to the
+`'a` parameter.  Were we to attempt passing `'a array` in a message, the
+compiler would be unable to prove that the message is compatible with the
+aforementioned constraints.  Therefore we must specify transitive constraints in
+the type, e.g. `'a <.  \& array`.  Similarly, finalization is only supported for
```

Comment:
In pretty much all the cases where we get a sneak peak at type specifier syntax, I skipped to the end of the doc to reference the meaning and wondered whether I'd be really confused had I not yet known the summary exists. I wonder if it'd be better to have the summary at the beginning. Maybe call it a "quick reference" if that's more appropriate since it no longer comes after the deeper explanations.

Reply:
Good idea; done.